bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/scanner/ResponseQueueReader.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright 2015 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.bigtable.grpc.scanner;
+
+import java.io.IOException;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+
+import com.google.bigtable.v1.ReadRowsResponse;
+import com.google.bigtable.v1.Row;
+import com.google.common.base.Optional;
+import com.google.common.base.Preconditions;
+
+/**
+ * Helper to read a queue of ResultQueueEntries and use the RowMergers to reconstruct
+ * complete Row objects from the partial ReadRowsResponse objects.
+ */
+class ResponseQueueReader {
+  private final BlockingQueue<ResultQueueEntry<ReadRowsResponse>> resultQueue;
+  private final int readPartialRowTimeoutMillis;
+  private boolean lastResponseProcessed = false;
+
+  public ResponseQueueReader(int capacity, int readPartialRowTimeoutMillis) {
+    this.resultQueue = new LinkedBlockingQueue<>(capacity);
+    this.readPartialRowTimeoutMillis = readPartialRowTimeoutMillis;
+  }
+
+  /**
+   * Get the next complete Row object from the response queue.
+   * @return Optional.absent() if end-of-stream, otherwise a complete Row.
+   * @throws IOException On errors.
+   */
+  public Optional<Row> getNextMergedRow() throws IOException {
```

Comment:
It wasn't before.  I made it synchronized.  I hope this doesn't affect performance too much.